src/swarm-tui/Swarm/TUI/View/Robot.hs

```diff
@@ -28,17 +42,223 @@ import Swarm.Game.State.Substate
+          , _cmdHistogramList = BL.list (RobotsListDialog $ SingleRobotDetails RobotCommandHistogramPane) mempty 1
+          }
+    }
+
+renderRobotsList :: RobotListContent -> Widget Name
+renderRobotsList rd =
+  vLimit 30 $
+    renderMixedTabularList (rd ^. robotsListRenderers) (LstFcs True) (rd ^. robotsListWidget)
+
+columnHdrAttr :: AttrName
+columnHdrAttr = attrName "columnHeader"
+
+rowHdrAttr :: AttrName
+rowHdrAttr = attrName "rowHeader"
+
+colHdr :: Set DebugOption -> MixedColHdr Name Widths
+colHdr uiDebug =
+  MixedColHdr
+    { draw = \_ (MColC (Ix ci)) -> case hdrs V.!? ci of
+        Just ch -> withAttr columnHdrAttr (str ch) <=> hBorder
+        Nothing -> emptyWidget
+    , widths = \(Widths ws) -> zipWith (<>) ws (map (ColW . length) $ V.toList hdrs)
+    , height = ColHdrH 2
+    }
+ where
+  hdrs = colHdrs uiDebug
+
+-- | Enumerates the rows by position (not 'RID').
+rowHdr :: RowHdr Name RobotWidgetRow
+rowHdr =
+  RowHdr
+    { draw = \_ (WdthD wd) (RowHdrCtxt (Sel s)) rh ->
+        let attrFn =
+              if s
+                then id
+                else withAttr rowHdrAttr
+         in attrFn $ padRight (Pad $ if wd > 0 then 0 else 1) $ padLeft Max (str $ show rh)
+    , width = \_ rh -> RowHdrW . (+ 2) . maximum0 $ map (length . show) rh
+    , toRH = \_ (Ix i) -> i + 1
+    }
+
+data ColumnExpansion
+  = Grow
+  | Minimal
+  deriving (Eq)
+
+data ColumnAttributes = Col
+  { headingString :: String
+  , expansionPolicy :: ColumnExpansion
+  }
+
+getColumnAttrList :: Set DebugOption -> NonEmpty ColumnAttributes
+getColumnAttrList dbgOptions =
+  NE.map ($ headingStrings) $ getAccessorList dbgOptions
+ where
+  headingStrings =
+    RobotRow
+      { _fID = Col "ID" Minimal
+      , _fName = Col "Name" Grow
+      , _fAge = Col "Age" Minimal
+      , _fPos = Col "Pos" Minimal
+      , _fItems = Col "Items" Minimal
+      , _fStatus = Col "Status" Minimal
+      , _fActns = Col "Actns" Minimal
+      , _fCmds = Col "Cmds" Minimal
+      , _fCycles = Col "Cycles" Minimal
+      , _fActivity = Col "Activity" Grow
+      , _fLog = Col "Log" Minimal
+      }
+
+colHdrs :: Set DebugOption -> Vector String
+colHdrs =
+  V.fromList
+    . NE.toList
+    . NE.map headingString
+    . getColumnAttrList
+
+getAccessorList :: Set DebugOption -> NonEmpty (RobotRow a -> a)
+getAccessorList dbgOptions =
+  applyWhen debugRID (NE.cons _fID) mainListSuffix
+ where
+  debugRID = dbgOptions ^. Lens.contains ListRobotIDs
+
+  mainListSuffix =
+    _fName
+      :| [ _fAge
+         , _fPos
+         , _fItems
+         , _fStatus
+         , _fActns
+         , _fCmds
+         , _fCycles
+         , _fActivity
+         , _fLog
+         ]
+
+drawCell :: Set DebugOption -> ListFocused -> MixedCtxt -> RobotWidgetRow -> Widget Name
+drawCell uiDebug _ (MxdCtxt _ (MColC (Ix ci))) r =
+  maybe emptyWidget (renderPlainCell . wWidget . ($ view rPayload r)) (indexedAccessors V.!? ci)
```

Comment:
What happens if the list is empty?

Does the tabular list require nonempty, or is that a self-imposed restriction?

Reply:
The `brick-tabular-list` package does not require nonempty column list, but I establish the invariant from the definition of `getAccessorList`.  That said, it probably would make sense for `brick-tabular-list` to have a `NonEmpty` restriction, since a table with zero columns doesn't make sense.